src/drivers/heater/heater.cpp

```diff
@@ -0,0 +1,439 @@
+ */
+
+#include "heater.h"
+
+#include <px4_getopt.h>
+#include <px4_log.h>
+#include <drivers/drv_hrt.h>
+
+#ifndef GPIO_HEATER_INPUT
+#error "To use the heater driver, the board_config.h must define and initialize GPIO_HEATER_INPUT and GPIO_HEATER_OUTPUT"
+#endif
+
+struct work_s Heater::_work = {};
+
+Heater::Heater() :
+	ModuleParams(nullptr),
+	_controller_period_usec(CONTROLLER_PERIOD_DEFAULT),
+	_controller_time_on_usec(0),
+	_duty_cycle(0.0f),
+	_heater_on(false),
+	_integrator_value(0.0f),
+	_proportional_value(0.0f),
+	_sensor_accel(sensor_accel_s{}),
+	_sensor_accel_sub(-1),
+	_sensor_temperature(0.0f)
+{
+	px4_arch_configgpio(GPIO_HEATER_OUTPUT);
+	px4_arch_gpiowrite(GPIO_HEATER_OUTPUT, 0);
+	_params_sub = orb_subscribe(ORB_ID(parameter_update));
+}
+
+Heater::~Heater()
+{
+	// Drive the heater GPIO pin low.
+	px4_arch_gpiowrite(GPIO_HEATER_OUTPUT, 0);
+
+	// Verify if GPIO is low, and if not, configure it as an input pulldown then reconfigure as an output.
+	if (px4_arch_gpioread(GPIO_HEATER_OUTPUT)) {
+		px4_arch_configgpio(GPIO_HEATER_INPUT);
+		px4_arch_configgpio(GPIO_HEATER_OUTPUT);
+		px4_arch_gpiowrite(GPIO_HEATER_OUTPUT, 0);
+	}
+}
+
+int Heater::controller_period(char *argv[])
+{
+	if (argv[1]) {
+		_controller_period_usec = atoi(argv[1]);
+	}
+
+	PX4_INFO("controller period (usec):  %i", _controller_period_usec);
+	return _controller_period_usec;
+}
+
+int Heater::custom_command(int argc, char *argv[])
+{
+	// Check if the driver is running.
+	if (!is_running() && !_object) {
+		PX4_INFO("not running");
+		return PX4_ERROR;
+	}
+
+	const char *arg_v = argv[0];
+
+	// Display/Set the heater controller period value (usec).
+	if (strcmp(arg_v, "controller_period") == 0) {
+		return get_instance()->controller_period(argv);
+	}
+
+	// Display the heater on duty cycle as a percent.
+	if (strcmp(arg_v, "duty_cycle") == 0) {
+		return get_instance()->duty_cycle();
+	}
+
+	// Display/Set the heater driver feed forward value.
+	if (strcmp(arg_v, "feed_forward") == 0) {
+		return get_instance()->feed_forward(argv);
+	}
+
+	// Display/Set the heater driver integrator gain value.
+	if (strcmp(arg_v, "integrator") == 0) {
+		return get_instance()->integrator(argv);
+	}
+
+	// Display/Set the heater driver proportional gain value.
+	if (strcmp(arg_v, "proportional") == 0) {
+		return get_instance()->proportional(argv);
+	}
+
+	// Display the id of the sensor we are controlling temperature on.
+	if (strcmp(arg_v, "sensor_id") == 0) {
+		return get_instance()->sensor_id();
+	}
+
+	// Displays/Set the current IMU temperature setpoint.
+	if (strcmp(arg_v, "setpoint") == 0) {
+		return get_instance()->temperature_setpoint(argv);
+	}
+
+	// Displays the IMU reported temperature.
+	if (strcmp(arg_v, "temp") == 0) {
+		return get_instance()->sensor_temperature();
+	}
+
+	get_instance()->print_usage("Unrecognized command.");
+	return PX4_OK;
+}
+
+void Heater::cycle()
+{
+	if (should_exit()) {
+		exit_and_cleanup();
```

Comment:
`exit_and_cleanup()` deletes the object, so the `orb_unsubscribe` need to be done before that. Can you do it in the destructor?
Also the printf is unnecessary.